internal/gps/constraints.go

```diff
@@ -52,6 +52,9 @@ type Constraint interface {
 	// It also forces Constraint to be a private/sealed interface, which is a
 	// design goal of the system.
 	typedString() string
+
+	// equals returns true if the constraints are logically equivalent.
```

Comment:
Per the other notes in this review, this should be clearer about the semantics of the equality check.